python_modules/dagster/dagster/_core/pipes/utils.py

```diff
@@ -362,12 +377,29 @@ def ext_asset(context: OpExecutionContext):
     context.set_requires_typed_event_stream(error_message=_FAIL_TO_YIELD_ERROR_MESSAGE)
     context_data = build_external_execution_context_data(context, extras)
     message_handler = PipesMessageHandler(context)
-    with context_injector.inject_context(
-        context_data
-    ) as ci_params, message_handler.handle_messages(message_reader) as mr_params:
-        yield PipesSession(
-            context_data=context_data,
-            message_handler=message_handler,
-            context_injector_params=ci_params,
-            message_reader_params=mr_params,
-        )
+    try:
+        with context_injector.inject_context(
+            context_data
+        ) as ci_params, message_handler.handle_messages(message_reader) as mr_params:
+            yield PipesSession(
+                context_data=context_data,
+                message_handler=message_handler,
+                context_injector_params=ci_params,
+                message_reader_params=mr_params,
+            )
+    finally:
+        if not message_handler.received_any_message:
+            context.log.warn(
+                "[pipes] did not receive any messages from external process. Check stdout / stderr"
+                " logs from the external process if"
+                f" possible.\n{context_injector.__class__.__name__}: "
+                f" {context_injector.no_messages_debug_text()}\n{message_reader.__class__.__name__}:"
+                f"  {message_reader.no_messages_debug_text()}\n"
+            )
+        elif not message_handler.received_closed_message:
+            context.log.warn(
+                "[pipes] did not receive closed message from external process. Buffered messages"
+                " may have been discarded without being delivered. Use `open_dagster_pipes` as a"
+                " context manager or explicitly call `PipesContext.close()` in the external"
+                " process to ensure delivery in most cases."
```

Comment:
It's subtle, but I would say something more like: 

Use `open_dagster_pipes` as a context manager (a with block) in the external process to ensure that cleanup is successfully completed. If using that is not possible, then manually call `PipesContext.close` in external process before it exists.